scripts/gen_binding.py

```diff
@@ -444,16 +481,16 @@ def normalize_arg(arg):
     js_impl_full = "\n".join(js_impl)
     wrap_args = list(zip(js_args, js_arg_types))
     if methods_only:
-        wrap_args = [('this', 'tensor')] + wrap_args
-    js_ptr_args = [x[0] if x[1] != "tensor" else f"{x[0]}.ptr" for x in wrap_args]
+        wrap_args = [('this', 'Tensor')] + wrap_args
+        js_tensor_args = ["this"] + js_tensor_args
+    js_ptr_args = [x[0] if x[1] != "Tensor" else f"{x[0]}.ptr" for x in wrap_args]
     js_ptr_result = f"const _ptr = fl._{op}({', '.join(js_ptr_args)})"
-    js_tensor_args = [f"{arg[0]}" for arg in wrap_args if arg[1] == "tensor"]
-    js_provenance_args = '||'.join([f"{t}.provenance" for t in js_tensor_args])
-    js_requires_grad_args = '||'.join([f"{t}.requires_grad" for t in js_tensor_args])
+    js_provenance_args = '||'.join([f"{t}.provenance" for t in js_tensor_args] + [f"{tv}.some((t) => t.provenance)" for tv in js_tensor_vector_args])
+    js_requires_grad_args = '||'.join([f"{t}.requires_grad" for t in js_tensor_args] + [f"{tv}.some((t) => t.requires_grad)" for tv in js_tensor_vector_args])
     js_requires_grad_args = 'false' if len(js_requires_grad_args) == 0 else js_requires_grad_args
-    js_requires_stats = '||'.join([f"{t}.requires_stats" for t in js_tensor_args])
+    js_requires_stats = '||'.join([f"{t}.requires_stats" for t in js_tensor_args] + [f"{tv}.some((t) => t.requires_stats)" for tv in js_tensor_vector_args])
     js_requires_stats = 'false' if len(js_requires_stats) == 0 else js_requires_stats
-    js_grad_args = (['this'] if methods_only else []) + js_args
+    js_grad_args = (['this'] if methods_only else []) + [f"...{n}" if t == "TensorVector" else f"{n}" for n, t in zip(js_grad_args, js_grad_arg_types)]
```

Comment:
I believe the logic previously was actually wrong.  `js_grad_args` should only be used for `js_deps` (to traverse backward) - which should really just be `Tensor`s.  Right now it includes `Shape`, `Axes` and other types.

It's not breaking, but I think we can actually clean up the differentiation code later if we can safely assume every dep is in fact a tensor.

Reply:
It seems to me that having the other types of arguments makes calculating gradient more straightforward, at least in the case of `concatenate`. Without it, we would have to add a bit of logic to infer the axis of concatenation based on the shapes of the input and output.

Reply:
ah you're totally right - I thought we were using a different way of passing inputs here: https://github.com/facebookresearch/shumai/blame/main/shumai/tensor/tensor.ts#L110

please disregard this comment :)